### ISSUE TITLE
Value inside parenthesized lambda is generalized when it shouldn't be

### ISSUE THREAD

**Issue by cartermp** [dotnet/fsharp#4144]
VS 2017, 15.5.2 + VSIX

Given the following code:

```fsharp
open System.Linq

[""].Join([""], (fun x -> x.
```

**Expected:** `x` is of type `string`, as the `TOuter` type [in the method](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.join?view=netframework-4.7) is a string.
**Actual:** `x` is of type `'a`

If I do not surround the lambda with parentheses, `x` is inferred to be `string`:

![oopsie](https://user-images.githubusercontent.com/6309070/34184181-3cd5e08e-e4d3-11e7-820d-4c90765b9e47.gif)

**Comment by cartermp** [comment]
I noticed this when trying to port over this test: https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/tests/unittests/LegacyLanguageService/Tests.LanguageService.Completion.fs#L443

It's hard to follow what this is precisely doing (as these tests are hard to follow in general), but it seems to me that it's definitely not testing what completion looks like _as you are typing the code out_.  If I copy the whole line, then I'll get the expected completion list that this tests.

This makes me feel that our existing, legacy test suite for completions is suspect :(

**Comment by majocha** [comment]
Alternatively, it's our completion that is not smart enough, and the tests reflect this.

If I close parenthesis like in the test, string type is properly inferred:
![image](https://user-images.githubusercontent.com/1760221/34185674-0871519e-e527-11e7-9b01-7f8caab24ee2.png)

**Comment by vasily-kirichenko** [comment]
Yes, this behavior has been there forever. I've got a habit to always close lambdas to get proper completion.

**Comment by cartermp** [comment]
As an update - for generated types, there is some other behavior where if I _add_ a `(` before a captured lambda value, it infers the type, but if I _don't_ have it added, it cannot infer it:

```fsharp
open FSharp.Data

[<Literal>]
let url = @"https://en.wikipedia.org/wiki/Megacity"

type MegaCities = HtmlProvider<url>

let printDuckFacts() =
    let data = MegaCities.GetSample()
    let tables = data.Tables
    
    let largestCities = tables.``Largest cities``

    let headers = largestCities.Headers

    let topFive =
        largestCities.Rows
        |> Array.take 5
        |> Array.map (fun x -> x. // <<-- x is inferred to be 'a

    printfn "Top 5 megacities: %A" topFive
```

<img width="190" alt="screen shot 2018-05-01 at 13 32 04" src="https://user-images.githubusercontent.com/6309070/39492316-10a67206-4d44-11e8-8075-e93ad894bfb3.png">

If I type this instead, I get IntelliSense (note the added `(`):

```fsharp
        |> Array.map (fun x -> (x.
```

<img width="310" alt="screen shot 2018-05-01 at 13 32 34" src="https://user-images.githubusercontent.com/6309070/39492361-2add7502-4d44-11e8-87dd-0ddb523361f2.png">

cc @dsyme

**Comment by cartermp** [comment]
This is significantly less of an issue as of automatic brace completion (which is on by default):
![better](https://user-images.githubusercontent.com/6309070/45902177-93f94400-bd9a-11e8-85ec-72c615415b87.gif)
